### PR TITLE
Move beacon code to a dedicated extension. Fix beacon remove vars

### LINKED ISSUE #3277: Error while submitting to beacon VCF files filtered by a gene panel
Example, [this case](https://scout-stage.scilifelab.se/cust002/F0038815) from cust002 in stage:

Filter variants using 2 gene panels:

image

Error:

image

### PR DESCRIPTION
**Pull request by northwestwitch**
- Fix #3277 
- Remove most of the code from cases views and controllers and put it into an app extension

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
</details>

**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

### CODE REVIEW

**Comment by northwestwitch** [comment]
Tested on stage and delete variants functionality now works:

image


Since I've tested with a case copied from prod and the panels for that case don't exist on stage (or don't exist in case dict?), I get an error when trying to create the submission data with the panel's genes:

image


Perhaps I can fix the code to fail a bit more gracefully..

**Comment by northwestwitch** [comment]
I think one of the problems is that the add variants endpoint times out when you provide one or more gene panels with many genes (probably the request is not even sent to the beacon). I'm investigating a solution (an async endpoint, using a thread?), but so far is still not working..

**Comment by northwestwitch** [comment]
Aha! Perhaps I don't need to complicate the code with async requests, there's a permission issue on the dockerized beacon app on stage: https://github.com/Clinical-Genomics/cgbeacon2/issues/227

**Comment by northwestwitch** [comment]
Another problem when you send requests with data containing thousands of genes:

image


I'm not sure this would work

**Comment by dnil** [comment]
> Another problem when you send requests with data containing thousands... 
> I'm not sure this would work

No, a say POST request with general data would be preferable to using the rather limited session object. I haven't had any time to read the code yet, so probably you already did! 😊 Let's check tomorrow!

**Comment by northwestwitch** [comment]
> > Another problem when you send requests with data containing thousands...
> > I'm not sure this would work
> 
> No, a say POST request with general data would be preferable to using the rather limited session object. I haven't had any time to read the code yet, so probably you already did! 😊 Let's check tomorrow!

Wait, it's not ready yet!

**Comment by northwestwitch** [comment]
> No, a say POST request with general data would be preferable to using the rather limited session object.

Agreed! We already have problems of this type with chanjo! I've been thinking of ways to "compress" a list of number in something that is smaller. Looks like Google has an algorithm that does exactly the same: https://developers.google.com/maps/documentation/utilities/polylinealgorithm

Check out this python implementation: https://github.com/amit1rrr/numcompress

**Comment by northwestwitch** [comment]
I think this is ready for review now!

File: scout/server/extensions/beacon_extension.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+            form(ImmutableMultiDict): request form submitted by user. Example:
+                [('case', 'internal_id'), ('samples', 'affected'), ('vcf_files', 'vcf_snv'), ('vcf_files', 'vcf_snv_research'), ('panels', '6246b25121d86882e127710c')]
+
+        """
+
+        # Check if user has rights to submit case to beacon
+        user_obj = store.user(current_user.email)
+
+        if "beacon_submitter" not in user_obj.get("roles", []):
+            flash("You don't have permission to use the Beacon tool", "warning")
```
Comment (dnil) [inline scout/server/extensions/beacon_extension.py:102]: Also something like please write to support and ask your institute primary contact to approve you as a Beacon submitter?

File: scout/server/extensions/beacon_extension.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        gene_ids = set()
+        for panel in form.getlist("panels"):
+            gene_ids.update(store.panel_to_genes(panel_id=panel, gene_format="hgnc_id"))
+
+        if gene_ids:
+            data["genes"] = {"ids": list(gene_ids), "id_type": "HGNC"}
+
+        return data
+
+    def add_variants(self, store, case_obj, form):
```
Comment (dnil) [inline scout/server/extensions/beacon_extension.py:87]: Passing the form all the way to the extension does connect things quite a bit. But the code is currently clean enough that I think changing it to parse the form in the view would make it more complex. All good for now!

File: tests/server/extensions/test_beacon_extension.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        # WHEN user submits variants from a case to the Beacon
+        beacon.add_variants(store, case_obj, None)
+
+        # THEN case will not be updated with submission data
+        updated_case = store.case_collection.find_one(case_obj)
+        assert "beacon" not in updated_case
+
+
+@responses.activate
+def test_add_variants_wrong_dataset(app, user_obj, case_obj):
```
Comment (dnil) [inline tests/server/extensions/test_beacon_extension.py:55]: 👍

**Review by dnil — APPROVED** [review]
Looks good to me, but lets test at stage as well.

File: scout/server/extensions/beacon_extension.py (hunk 1 above)
Comment (northwestwitch) [reply scout/server/extensions/beacon_extension.py:102]: Yes, good idea!

File: scout/server/extensions/beacon_extension.py (hunk 2 above)
Comment (northwestwitch) [reply scout/server/extensions/beacon_extension.py:87]: Did you mean that it is fine to pass the raw request of it should be parsed in the view first? Sorry i didn't get it

Comment (dnil) [reply scout/server/extensions/beacon_extension.py:87]: I think it is in principle more correct if the view parses requests - or possibly a dedicated function from the controller. But we don’t hold to that in general, and this code is still very clean - the forms are not touched until they reach the extension, so I do not mind an exception!

Comment (northwestwitch) [reply scout/server/extensions/beacon_extension.py:87]: Ah OK, thanks!

**Comment by northwestwitch** [comment]
Test trying to submit a case to a dataset that doesn't exist in beacon: 🆗 

image

Test add variants to dataset that exists (cust002 GRCH37) 🆗 

image

Test removing vars for the same case: 🆗 

image

**Comment by northwestwitch** [comment]
Merging this one as well!

**Pull request merged**